metadetect/metadetect.py

```diff
@@ -419,29 +568,60 @@ def _add_positions_and_psf(self, cat, res, shear_str):
 
         return newres
 
-    def _do_detect(self, mbobs):
+    def _do_detect(self, mbobs, shear_bands):
         """
         use a MEDSifier to run detection
         """
-        return detect.MEDSifier(
-            mbobs=mbobs,
+        t0 = time.time()
+        shear_mbobs = ngmix.MultiBandObsList()
+        for band in shear_bands:
+            shear_mbobs.append(mbobs[band])
+
+        medsifier = detect.MEDSifier(
+            mbobs=shear_mbobs,
             sx_config=self['sx'],
             meds_config=self['meds'],
             nodet_flags=self['nodet_flags'],
         )
 
+        if self._show:
+            import descwl_coadd.vis
+            descwl_coadd.vis.show_image(medsifier.seg)
+
+        all_medsifier = detect.CatalogMEDSifier(
+            mbobs,
+            medsifier.cat['x'],
+            medsifier.cat['y'],
+            medsifier.cat['box_size'],
+        )
+        mbm = all_medsifier.get_multiband_meds()
+        mbobs_list = mbm.get_mbobs_list()
+        logger.info("detect took %s seconds", time.time() - t0)
+
+        return medsifier.cat, mbobs_list
+
     def _get_all_metacal(self, mbobs):
         """
         get the sheared versions of the observations
         """
+        # we cache this locally - may not be the best idea but here we are
+        if "__mdet_mcal_res" not in mbobs.meta:
+            t0 = time.time()
+            try:
+                odict = ngmix.metacal.get_all_metacal(
+                    mbobs,
+                    rng=self.rng,
+                    **self['metacal']
+                )
+            except BootPSFFailure:
+                odict = None
+            logger.info("metacal took %s seconds", time.time() - t0)
 
-        odict = ngmix.metacal.get_all_metacal(
-            mbobs,
-            rng=self.rng,
-            **self['metacal']
-        )
+            mbobs.meta["__mdet_mcal_res"] = odict
```

Comment:
Is there any danger of mbobs changing on successive calls?  If so, might want to cache in a dict indexed on mbobs or check that mbobs matches what was used for the cache result.